src/cumulativeSum.js

```diff
@@ -19,20 +19,23 @@ export function cumulativeSum(input, axis, {exclusive = false, reverse = false}
   const totalElements = sizeOfShape(outputShape);
   const inputElementStart = reverse ? elementCountAlongAxis - 1 : 0;
   const inputElementStep = reverse ? -1 : 1;
+
   const cumulativeSums = new Array(elementCountAlongAxis).fill(0);
+
   for (let outputIndex = 0; outputIndex < totalElements; outputIndex++) {
     const location = output.locationFromIndex(outputIndex);
-    const inputLocation = [...location];
-    const outputLocation = [...location];
+
+    if (location[axis] !== inputElementStart) continue;
+
     for (let i = 0; i < elementCountAlongAxis; ++i) {
-      const idx = inputElementStart + i * inputElementStep;
-      inputLocation[axis]=idx;
-      outputLocation[axis] = idx;
-      const inputValue = input.getValueByLocation(inputLocation);
-      cumulativeSums[i] = (i === 0 ? 0 : cumulativeSums[i-1]) + inputValue;
-      const outputValue = exclusive ? (i === 0 ? 0 : cumulativeSums[i-1]) : cumulativeSums[i];
-      output.setValueByLocation(outputLocation, outputValue);
+      const index = inputElementStart + i * inputElementStep;
+      location[axis] = index;
+      const inputValue = input.getValueByLocation(location);
+      cumulativeSums[i] = (i === 0 ? 0 : cumulativeSums[i - 1]) + inputValue;
+      const outputValue = exclusive ? (i === 0 ? 0 : cumulativeSums[i - 1]) : cumulativeSums[i];
```

Comment:
Do we need a full array of values rather than just the running tally? Maybe you were using it for debugging to inspect values, but it feels it would be easier for future readers to follow something simpler like:

```
const cumulativeSum = exclusive ? previousSum : previousSum + inputValue;
previousSum += inputValue;
```

Additionally the current logic is kinda repetitive, stating the identical `(i === 0 ? 0 : cumulativeSums[i - 1]) ...` part twice, and you have to worry about special cases like the -1 case.

Reply:
Thank you for your advice! :)